From 3229945a2f3dbb13e1d9ff724e04a36dffb00231 Mon Sep 17 00:00:00 2001
From: Johnny Wang <wangjiahao@openresty.com>
Date: Tue, 16 Nov 2021 18:01:07 +0800
Subject: [PATCH] bugfix: prevent illegal memory access in ngx_http_lua_util.c.
 (#1974)

found by core dump caused by running TEST 3 of t/072-conditional-get.t
in mockeagain write mode.
---
 src/ngx_http_lua_util.c | 12 ++++++++++--
 1 file changed, 10 insertions(+), 2 deletions(-)

diff --git a/ngx_lua-0.10.20/src/ngx_http_lua_util.c b/ngx_lua-0.10.20/src/ngx_http_lua_util.c
index 0d29c7bcb..bae59daf3 100644
--- a/ngx_lua-0.10.20/src/ngx_http_lua_util.c
+++ b/ngx_lua-0.10.20/src/ngx_http_lua_util.c
@@ -719,6 +719,10 @@ ngx_http_lua_output_filter(ngx_http_request_t *r, ngx_chain_t *in)
 
     ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
 
+    if (ctx == NULL) {
+        return rc;
+    }
+
     ngx_chain_update_chains(r->pool,
                             &ctx->free_bufs, &ctx->busy_bufs, &in,
                             (ngx_buf_tag_t) &ngx_http_lua_module);
@@ -1027,10 +1031,14 @@ ngx_http_lua_generic_phase_post_read(ngx_http_request_t *r)
 
     ctx = ngx_http_get_module_ctx(r, ngx_http_lua_module);
 
-    ctx->read_body_done = 1;
-
     r->main->count--;
 
+    if (ctx == NULL) {
+        return;
+    }
+
+    ctx->read_body_done = 1;
+
     if (ctx->waiting_more_body) {
         ctx->waiting_more_body = 0;
         ngx_http_core_run_phases(r);
